framework/platform/plugins/parser.cpp

```diff
@@ -0,0 +1,396 @@
+/* Copyright (c) 2019-2021, Arm Limited and Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "parser.h"
+
+#include <iostream>
+#include <set>
+#include <sstream>
+
+#include <spdlog/fmt/fmt.h>
+
+#include "common/logging.h"
+#include "common/strings.h"
+#include "platform/plugins/plugin.h"
+
+namespace vkb
+{
+Parser::Parser(const std::vector<Plugin *> &plugins)
+{
+	std::vector<Plugin *> entrypoints     = plugins::with_tags<tags::Entrypoint>(plugins);
+	std::vector<Plugin *> not_entrypoints = plugins::without_tags<tags::Entrypoint>(plugins);
+
+	// Dont mix well togther
+	std::vector<Plugin *> aggressive = plugins::with_tags<tags::FullControl, tags::Stopping>(not_entrypoints);
+
+	// Work well with any plugin
+	std::vector<Plugin *> passives = plugins::with_tags<tags::Passive>(not_entrypoints);
+
+	// Entrypoint - {aggressive, passive}
+	std::unordered_map<Plugin *, std::vector<Plugin *>> usage;
+	usage.reserve(entrypoints.size());
+
+	for (auto entrypoint : entrypoints)
+	{
+		std::vector<Plugin *> compatable;
+
+		if (!entrypoint->has_tag<tags::FullControl>() || entrypoint->has_tag<tags::Stopping>())
+		{
+			// The entrypoint does not dictate the applications functionality so allow other plugins to take control
+			compatable.reserve(compatable.size() + aggressive.size());
+			for (auto ext : aggressive)
+			{
+				compatable.push_back(ext);
+			}
+		}
+
+		compatable.reserve(compatable.size() + passives.size());
+		for (auto ext : passives)
+		{
+			compatable.push_back(ext);
+		}
+
+		usage.insert({entrypoint, compatable});
+	}
+
+	const char *spacer   = "  ";
+	const char *app_name = "VulkanSamples";
```

Comment:
The app name is actually "vulkan_samples". This causes all the usage to display the wrong command.

Reply:
Fixed